pkg/cri/resource-manager/policy/builtin/memtier/resources.go

```diff
@@ -460,39 +496,63 @@ func (cs *supply) Allocate(r Request) (Grant, error) {
 
 	cr := r.(*request)
 
-	// allocate isolated exclusive CPUs or slice them off the sharable set
-	switch {
-	case cr.full > 0 && cs.isolated.Size() >= cr.full && cr.isolate:
-		exclusive, err = cs.takeCPUs(&cs.isolated, nil, cr.full)
-		if err != nil {
-			return nil, policyError("internal error: "+
-				"%s: can't take %d exclusive isolated CPUs from %s: %v",
-				cs.node.Name(), cr.full, cs.isolated, err)
-		}
+	full := cr.full
+	fraction := cr.fraction
+	if cr.cpuType == cpuNormal {
+		// allocate isolated exclusive CPUs or slice them off the sharable set
+		switch {
+		case full > 0 && cs.isolated.Size() >= full && cr.isolate:
+			exclusive, err = cs.takeCPUs(&cs.isolated, nil, full)
+			if err != nil {
+				return nil, policyError("internal error: "+
+					"%s: can't take %d exclusive isolated CPUs from %s: %v",
+					cs.node.Name(), full, cs.isolated, err)
+			}
 
-	case cr.full > 0 && cs.AllocatableSharedCPU() > 1000*cr.full:
-		exclusive, err = cs.takeCPUs(&cs.sharable, nil, cr.full)
-		if err != nil {
+		case full > 0 && cs.AllocatableSharedCPU() > 1000*full:
+			exclusive, err = cs.takeCPUs(&cs.sharable, nil, full)
+			if err != nil {
+				return nil, policyError("internal error: "+
+					"%s: can't take %d exclusive CPUs from %s: %v",
+					cs.node.Name(), full, cs.sharable, err)
+			}
+
+		case full > 0:
 			return nil, policyError("internal error: "+
-				"%s: can't take %d exclusive CPUs from %s: %v",
-				cs.node.Name(), cr.full, cs.sharable, err)
+				"%s: can't slice %d exclusive CPUs from %s, %dm available",
+				cs.node.Name(), full, cs.sharable, cs.AllocatableSharedCPU())
+		}
+	} else if cr.cpuType == cpuReserved {
+		if full > 0 {
+			// no support for isolated allocation of ReservedCPUs
+			fraction += full * 1000
+			full = 0
```

Comment:
That is correct. This was a piece of defensive coding. For instance, if one day something somewhere would rebalance or otherwise modify a request or a grant and call `SetCPUType(cpuReserved)` for containers in certain namespace. Or add new cpuAllocationPreference that would return full > 0 for reserved CPUs.

Would you prefer having here a warning or dropping this branch completely?

Reply:
Fair enough. We can leave it like that.